Locking incorrectly claims that dependency is not installed

I might be on a spree of opam locking issues this week...

At https://github.com/sim642/odep/tree/67ce57a013f342da49cdbe73025130d66cd06e12, opam 2.5.1 gives the following message:
```console
$ opam lock .
[ERROR] Skipping odep.0.2.1, dependencies are not satisfied in this switch, not installed packages are:
          - opam-state (< 2.2.0~beta3)

No lock file generated
```

However, the opam switch I'm in is exported here: https://github.com/sim642/odep/blob/67ce57a013f342da49cdbe73025130d66cd06e12/odep.opam.exported#L79. It has opam-state 2.5.1 installed but `opam lock` somehow doesn't see this constraint in its entirety:
```
"opam-state" {>= "2.1.0" & (< "2.2.0~beta3" | >= "2.2.0")}
```


---


I thought that maybe it's still trying to lock based on the constraints of odep.0.2.1 in opam-repository, so I did `opam pin odep.0.2.2 .`.
This installed fine (opam-state was already installed), but `opam lock .` claims the same (except for odep.0.2.2 this time).

Thanks for the report, i was indeed able to reproduce this (using our reftests syntax):
```
### <pkg:dep.1>
opam-version: "2.0"
### <pkg:dep.2>
opam-version: "2.0"
### <pin:disjunction/disjunction.opam>
opam-version: "2.0"
depends: ["dep" {= "1" | = "2"}]
### opam install dep.2
The following actions will be performed:
=== install 1 package
  - install dep 2

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> installed dep.2
Done.
### opam lock disjunction/
[ERROR] Skipping disjunction.dev, dependencies are not satisfied in this switch, not installed packages are:
          - dep (= 1)

No lock file generated
```
For some reason `opam lock` seems to simply ignore disjunctions

--- a/master_changes.md
+++ b/master_changes.md
@@ -86,6 +86,7 @@ users)
 
 ## Lock
   * [BUG] Fix `undefined variable` error when a lock file filter contains an undefined variables: fail gracefully with strict mode, continue and default the variable to false on normal mode [#6947 @rjbou - fix #6946]
+  * [BUG] Fix `opam lock` support of dependency formula that include disjunctions [#6990 @kit-ty-kate - fix #6944]
 
 ## Clean
 

--- a/master_changes.md
+++ b/master_changes.md
@@ -232,6 +232,7 @@ users)
   * Add a test showing the behaviour of `opam init` when the default compiler failed to build [#6851 @kit-ty-kate]
   * Add a test ensuring `/` can be used as directory separator on Windows when referring to a path to a local repository [#6981 @kit-ty-kate]
   * Add a test showing when opam creates a local switch vs. global [#6981 @kit-ty-kate]
+  * Add a test showing the behaviour of `opam lock` in presence of disjunction in the dependency formula [#6990 @kit-ty-kate]
 
 ### Engine
   * Add `http-server` to launch a minimal http server [#6939 @rjbou]

--- a/src/client/opamLockCommand.ml
+++ b/src/client/opamLockCommand.ml
@@ -80,27 +80,28 @@ let select_packages atom_locs st =
                | None -> OpamSwitchState.opam st nv
              else OpamSwitchState.opam st nv
            in
-           let atoms =
-             OpamFormula.atoms
+           let cnf =
+             OpamFormula.to_cnf
                (OpamPackageVar.all_depends ~depopts:false st opam)
            in
            let missing =
-             List.filter (fun (n,vc) ->
-                 let pkgs =
-                   OpamPackage.packages_of_name
-                     (OpamPackage.Set.union st.installed packages)
-                     n
-                 in
-                 OpamPackage.Set.is_empty pkgs ||
-                 not (OpamPackage.Set.exists (OpamFormula.check (n,vc)) pkgs))
-               atoms
+             let installed_pkgs = OpamPackage.Set.union st.installed packages in
+             List.filter (fun disj ->
+                 List.for_all (fun (n,vc) ->
+                     let pkgs = OpamPackage.packages_of_name installed_pkgs n in
+                     OpamPackage.Set.is_empty pkgs ||
+                     not (OpamPackage.Set.exists (OpamFormula.check (n,vc)) pkgs))
+                   disj)
+               cnf
            in
            if missing <> [] then
              (OpamConsole.error
                 "Skipping %s, dependencies are not satisfied in this switch, \
                  not installed packages are:\n%s"
                 (OpamPackage.to_string nv)
-                (OpamStd.Format.itemize OpamFormula.string_of_atom missing);
+                (OpamStd.Format.itemize
+                   (OpamStd.List.concat_map " | " OpamFormula.string_of_atom)
+                   missing);
               acc)
            else
              OpamPackage.Set.add nv acc)

--- a/tests/reftests/lock.test
+++ b/tests/reftests/lock.test
@@ -1400,3 +1400,32 @@ with-nope is now pinned to file://${BASEDIR}/with-nope (version dev)
 [ERROR] Undefined filter variable with-nope in dependencies of with-nope.dev
 [NOTE] Pinning command successful, but your installed packages may be out of sync.
 # Return code 30 #
+### opam unpin with-nope
+Ok, with-nope is no longer pinned to file://${BASEDIR}/with-nope (version dev)
+### : disjunctions
+### <pkg:dep.1>
+opam-version: "2.0"
+### <pkg:dep.2>
+opam-version: "2.0"
+### <pin:disjunction/disjunction.opam>
+opam-version: "2.0"
+depends: ["dep" {= "1" | = "2"}]
+### opam lock disjunction/
+[ERROR] Skipping disjunction.dev, dependencies are not satisfied in this switch, not installed packages are:
+          - dep (= 1)
+          - dep (= 2)
+
+No lock file generated
+### opam install dep.2
+The following actions will be performed:
+=== install 1 package
+  - install dep 2
+
+<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
+-> installed dep.2
+Done.
+### opam lock disjunction/
+[ERROR] Skipping disjunction.dev, dependencies are not satisfied in this switch, not installed packages are:
+          - dep (= 1)
+
+No lock file generated

--- a/tests/reftests/lock.test
+++ b/tests/reftests/lock.test
@@ -108,17 +108,17 @@ depends:   ["bar" {= "1"} "baz" {= "1"} "foo" {= "1"}]
 depopts:
 conflicts:
 ### : error or formulaes :
-### ### ERROR ###
 ### <pin:tolock/tolock.opam>
 opam-version: "2.0"
 depends: [ "foo" ("bar" | "qux") ]
 depopts: "baz"
 build: [ "test" "-f" "content" ]
+### opam list -s --installed foo bar qux
+bar
+foo
 ### opam lock tolock
-[ERROR] Skipping tolock.1, dependencies are not satisfied in this switch, not installed packages are:
-          - qux
-
-No lock file generated
+Generated lock files for:
+  - tolock.1: ${BASEDIR}/tolock.opam.locked
 ### opam show --just-file --normalise ./tolock.opam.locked --field depends,depopts,conflicts
 depends:   ["bar" {= "1"} "baz" {= "1"} "foo" {= "1"}]
 depopts:
@@ -1412,8 +1412,7 @@ opam-version: "2.0"
 depends: ["dep" {= "1" | = "2"}]
 ### opam lock disjunction/
 [ERROR] Skipping disjunction.dev, dependencies are not satisfied in this switch, not installed packages are:
-          - dep (= 1)
-          - dep (= 2)
+          - dep (= 1) | dep (= 2)
 
 No lock file generated
 ### opam install dep.2
@@ -1425,7 +1424,5 @@ The following actions will be performed:
 -> installed dep.2
 Done.
 ### opam lock disjunction/
-[ERROR] Skipping disjunction.dev, dependencies are not satisfied in this switch, not installed packages are:
-          - dep (= 1)
-
-No lock file generated
+Generated lock files for:
+  - disjunction.dev: ${BASEDIR}/disjunction.opam.locked

